azurelinuxagent/distro/suse/rdma.py

```diff
@@ -48,22 +51,42 @@ def install_driver(self):
                     continue
                 installed = sections[0].strip()
                 version = sections[3].strip()
-                if fw_version in version:
-                    if installed == 'i':
-                        info_msg = 'Driver package "%s-%s" '
-                        info_msg += 'is already installed, nothing to do'
-                        logger.info(info_msg % (package_name, version))
-                        break
-                    cmd = '%s %s-%s' % (zypper_install, package_name, version)
-                    result = os.system(cmd)
-                    if result:
-                        error_msg = 'Failed install of package "%s-%s" '
-                        error_msg += 'from available repositories.'
-                        logger.error(error_msg % (package_name, version))
-                    msg = 'Successfully installed "%s-%s" from '
-                    msg += 'configured repositories'
-                    logger.info(msg % (package_name, version))
-                    break
+                driver_package_versions.append(version)
+                if fw_version in version and installed == 'i':
+                    info_msg = 'Matching driver package "%s-%s" '
+                    info_msg += 'is already installed, nothing to do'
+                    logger.info(info_msg % (package_name, version))
+                    return True
+                if installed == 'i':
+                    driver_package_installed = True
+
+        # If we get here the driver package is installed but the
+        # version doesn't match or no package is installed
+        requires_reboot = False
+        if driver_package_installed:
+            # Unloading the particular driver with rmmod does not work
+            # We have to reboot after the new driver is installed
+            if self.is_driver_loaded():
+                requires_reboot = True
+            cmd = zypper_remove % package_name
+            shellutil.run(cmd)
+
+        for entry in driver_package_versions:
+            if fw_version in version:
+                complete_name = '%s-%s' % (package_name, version)
+                cmd = zypper_install % complete_name
+                result = shellutil.run(cmd)
+                if result:
+                    error_msg = 'Failed install of package "%s" '
+                    error_msg += 'from available repositories.'
+                    logger.error(error_msg % complete_name)
+                msg = 'Successfully installed "%s" from '
+                msg += 'configured repositories'
+                logger.info(msg % complete_name)
+                self.load_driver_module()
+                if requires_reboot:
```

Comment:
can we please add some logging around the reboot. it is quite a critical operation for us?